File: Part7/C++/Maze_N_x_N/Maze_N_x_N.cpp
#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <thread>

int main ( )
{
    // Define directions
    //const int LEFT = 0;
    //const int UP = 1;
    //const int RIGHT = 2;
    //const int DOWN = 3;

    const int FIRST_CELL = 0;
    const int SECOND_CELL = 1;
    
    const int NO_CELL = -1;
    
    // Prompt the user for maze size
    int mazeColumns = 0;
    while ( mazeColumns <= 0 )
    {
        printf ( "Please enter number of columns for maze, must be greater than 1: " );
        char userInput[100];
        if ( fgets ( userInput, sizeof ( userInput ), stdin ) != nullptr )
        {
            // Remove newline if present
            userInput[strcspn ( userInput, "\n" )] = '\0';
            
            // Check if string contains only digits
            bool isValid = true;
            if ( strlen ( userInput ) == 0 )
            {
                isValid = false;
            }
            else
            {
                for ( int i = 0; userInput[i] != '\0'; i++ )
                {
                    if ( userInput[i] < '0' || userInput[i] > '9' )
                    {
                        isValid = false;
                        break;
                    }
                }
            }
            
            if ( isValid )
            {
                mazeColumns = atoi ( userInput );
            }
        }
    }

    int mazeRows = 0;
    while ( mazeRows <= 0 )
    {
        printf ( "Please enter number of rows for maze, must be greater than 1: " );
        char userInput[100];
        if ( fgets ( userInput, sizeof ( userInput ), stdin ) != nullptr )
        {
            // Remove newline if present
            userInput[strcspn ( userInput, "\n" )] = '\0';
            
            // Check if string contains only digits
            bool isValid = true;
            if ( strlen ( userInput ) == 0 )
            {
                isValid = false;
            }
            else
            {
                for ( int i = 0; userInput[i] != '\0'; i++ )
                {
                    if ( userInput[i] < '0' || userInput[i] > '9' )
                    {
                        isValid = false;
                        break;
                    }
                }
            }
            
            if ( isValid )
            {
                mazeRows = atoi ( userInput );
            }
        }
    }

    int interiorWallCount = mazeRows * ( mazeColumns - 1 ) + ( mazeRows - 1 ) * mazeColumns;

    // Start with all the walls up.
    bool wallsUp[interiorWallCount];
    for ( int i = 0; i < interiorWallCount; i++ )
    {
        wallsUp[i] = true;
    }

    // Identify the cells each wall connects.
    int wallConnections[interiorWallCount][2];

    int wallIndex = 0;
    for ( int rowIndex = 0; rowIndex <= mazeRows - 1; rowIndex++ )
    {
        // Track the first cell in the current row
        int firstCellInRow = rowIndex * mazeColumns;

        // Note the 0..mazeColumns - 2, one less vertical wall
        // than the number of columns.
        for ( int verticalWallIndex = 0; verticalWallIndex <= mazeColumns - 2; verticalWallIndex++ )
        {
            int leftCell = firstCellInRow + verticalWallIndex;
            int rightCell = leftCell + 1;
            wallConnections[wallIndex][FIRST_CELL] = leftCell;
            wallConnections[wallIndex][SECOND_CELL] = rightCell;
            wallIndex = wallIndex + 1;
        }

        // The last row will have no interior horizontal walls below
        // it, so will be skipped.
        if ( wallIndex < interiorWallCount )
        {
            for ( int horizontalWallIndex = 0; horizontalWallIndex <= mazeColumns - 1; horizontalWallIndex++ )
            {
                int upperCell = firstCellInRow + horizontalWallIndex;
                int lowerCell = upperCell + mazeColumns;
                wallConnections[wallIndex][FIRST_CELL] = upperCell;
                wallConnections[wallIndex][SECOND_CELL] = lowerCell;
                wallIndex = wallIndex + 1;
            }
        }
    }

    // Identify which group each cell is a part of.
    int cellToGroup[mazeRows * mazeColumns];

    for ( int cellIndex = 0; cellIndex <= mazeRows * mazeColumns - 1; cellIndex++ )
    {
        cellToGroup[cellIndex] = cellIndex;
    }

    // Identify which cells are a part of each group
    // Note each array is large enough to fit all cells,
    // NO_CELL indicates no cell is assigned to this index.
    int groupCells[mazeColumns * mazeRows][mazeColumns * mazeRows];
    
    for ( int cellIndex = 0; cellIndex <= mazeRows * mazeColumns - 1; cellIndex++ )
    {
        groupCells[cellIndex][0] = cellIndex;
    }

    // Print maze code:
    // Print out the maze, this is a less painful copy/paste job without functions, but better with loops.
    int currentInteriorWall = 0;

    // Print the horizontal walls above row 1 - All are exterior walls, no conditions.
    // +-+-+-+
    // One initial cell with + followed by mazeColumns cells with -+
    printf ( "+" );
    for ( int cellIndex = 0; cellIndex <= mazeColumns - 1; cellIndex++ )
    {
        printf ( "-" );
        printf ( "+" );
    }
    printf ( "\n" );

    for ( int rowIndex = 0; rowIndex <= mazeRows - 1; rowIndex++ )
    {
        // Vertical walls and cells in each row row.
        // The left and right vertical walls are exterior, always up.
        // | | | |
        // Or print one |, followed by mazeColumn cells of <space>| where the |
        // may be down (<space>).
        printf ( "|" );
        for ( int columnIndex = 0; columnIndex <= mazeColumns - 1; columnIndex++ )
        {
            printf ( " " );

            // Always print the right most vertical wall,
            // if interior wall, print if the wall is up.
            if ( columnIndex == mazeColumns - 1 || wallsUp[currentInteriorWall] == true )
            {
                printf ( "|" );
            }
            else
            {
                printf ( " " );
            }
            if ( columnIndex < mazeColumns - 1 )
            {
                currentInteriorWall = currentInteriorWall + 1;
            }
        }
        printf ( "\n" );

        // One fewer horizontal wall than vertical
        if ( rowIndex < mazeRows - 1 )
        {
            // Horizontal walls above row rowIndex
            // +-+-+-+
            printf ( "+" );
            for ( int columnIndex = 0; columnIndex <= mazeColumns - 1; columnIndex++ )
            {
                if ( wallsUp[currentInteriorWall] == true )
                {
                    printf ( "-" );
                }
                else
                {
                    printf ( " " );
                }
                printf ( "+" );
                currentInteriorWall = currentInteriorWall + 1;
            }
            printf ( "\n" );
        }
    }

    // Horizontal walls below row the final row - All are exterior walls, no conditions.
    // +-+-+-+
    printf ( "+" );
    for ( int columnIndex = 0; columnIndex <= mazeColumns - 1; columnIndex++ )
    {
        printf ( "-" );
        printf ( "+" );
    }
    printf ( "\n" );

    int wallRemoveList[interiorWallCount];
    for ( int wallIndex = 0; wallIndex <= interiorWallCount - 1; wallIndex++ )
    {
        wallRemoveList[wallIndex] = wallIndex;
    }

    srand ( time ( NULL ) );

    // Fisher-Yates shuffle algorithm
    for ( int shuffleIndex = interiorWallCount - 1; shuffleIndex > 0; shuffleIndex = shuffleIndex - 1 )
    {
        // Generate random index from 0 to shuffleIndex (inclusive)
        int otherIndex = rand ( ) % ( shuffleIndex + 1 );

        // Swap wallRemoveList[shuffleIndex] with wallRemoveList[otherIndex]
        int temp = wallRemoveList[shuffleIndex];
        wallRemoveList[shuffleIndex] = wallRemoveList[otherIndex];
        wallRemoveList[otherIndex] = temp;
    }

    bool mazeComplete = false;

    // Remove wall code:
    // Now that we have loops we can implement Kruskal's algorithm.
    // The simple description of the algorithm is first place each 
    // cell in its own group.  Then process all walls in random order,
    // if the cells on either side of the wall are in separate groups, 
    // remove the wall and merge the groups.  Repeat until all 
    // cells are now in the same group.
    for ( int removeWallIndex = 0; removeWallIndex <= interiorWallCount - 1; removeWallIndex++ )
    {
        int nextWallToCheck = wallRemoveList[removeWallIndex];

        // If the two cells connected to this wall are not part 
        // of the same group, remove the wall and merge the 
        // groups.
        int firstCell = wallConnections[nextWallToCheck][FIRST_CELL];
        int firstCellGroupIndex = cellToGroup[firstCell];
        int secondCell = wallConnections[nextWallToCheck][SECOND_CELL];
        int secondCellGroupIndex = cellToGroup[secondCell];
        
        if ( firstCellGroupIndex != secondCellGroupIndex )
        {
            wallsUp[nextWallToCheck] = false;

            // Loop through the indices of all cells in the first 
            // group until we find a NO_CELL indicating no cell here.
            int nextEmptyFirstGroupIndex = 0;
            for ( int cellIndex = 0; cellIndex <= mazeColumns * mazeRows - 1; cellIndex++ )
            {
                if ( groupCells[firstCellGroupIndex][cellIndex] == NO_CELL )
                {
                    nextEmptyFirstGroupIndex = cellIndex;
                    break;
                }
            }

            // Loop through the indices of all cells in the second group,
            // move each cell to the first group, and set that cell's 
            // group to the first group index.
            for ( int groupCellIndex = mazeColumns * mazeRows - 1; groupCellIndex >= 0; groupCellIndex-- )
            {
                // Skip until we reach valid cells
                if ( groupCells[secondCellGroupIndex][groupCellIndex] != NO_CELL )
                {
                    // Get the id number of the cell to move from 
                    // the second group to the first group
                    int cellToMove = groupCells[secondCellGroupIndex][groupCellIndex];

                    // Move the cell number from the second group 
                    // to the first group
                    groupCells[firstCellGroupIndex][nextEmptyFirstGroupIndex] = cellToMove;
                    // Move our empty index to the next cell in this array.
                    nextEmptyFirstGroupIndex = nextEmptyFirstGroupIndex + 1;
                    // Mark this cell as part of the first group.
                    cellToGroup[cellToMove] = firstCellGroupIndex;
                    // Remove the cell from the second group (set the
                    // array entry to NO_CELL)
                    groupCells[secondCellGroupIndex][groupCellIndex] = NO_CELL;
                    
                    if ( nextEmptyFirstGroupIndex >= mazeColumns * mazeRows )
                    {
                        mazeComplete = true;
                    }
                }
            }

            std::this_thread::sleep_for ( std::chrono::milliseconds ( 500 ) );
                    
            // Copy in Print maze code from above again here:
            currentInteriorWall = 0;

            printf ( "\n" );
            printf ( "+" );
            for ( int cellIndex = 0; cellIndex <= mazeColumns - 1; cellIndex++ )
            {
                printf ( "-" );
                printf ( "+" );
            }
            printf ( "\n" );

            for ( int rowIndex = 0; rowIndex <= mazeRows - 1; rowIndex++ )
            {
                printf ( "|" );
                for ( int columnIndex = 0; columnIndex <= mazeColumns - 1; columnIndex++ )
                {
                    printf ( " " );

                    if ( columnIndex == mazeColumns - 1 || wallsUp[currentInteriorWall] == true )
                    {
                        printf ( "|" );
                    }
                    else
                    {
                        printf ( " " );
                    }
                    if ( columnIndex < mazeColumns - 1 )
                    {
                        currentInteriorWall = currentInteriorWall + 1;
                    }
                }
                printf ( "\n" );

                if ( rowIndex < mazeRows - 1 )
                {
                    printf ( "+" );
                    for ( int columnIndex = 0; columnIndex <= mazeColumns - 1; columnIndex++ )
                    {
                        if ( wallsUp[currentInteriorWall] == true )
                        {
                            printf ( "-" );
                        }
                        else
                        {
                            printf ( " " );
                        }
                        printf ( "+" );
                        currentInteriorWall = currentInteriorWall + 1;
                    }
                    printf ( "\n" );
                }
            }

            printf ( "+" );
            for ( int columnIndex = 0; columnIndex <= mazeColumns - 1; columnIndex++ )
            {
                printf ( "-" );
                printf ( "+" );
            }
            printf ( "\n" );

            if ( mazeComplete == true )
            {
                break;
            }
        }
    }

    return 0;
}

File: Part7/Fortran/Maze_N_x_N/Maze_N_x_N.f90
program mazeGenerator
    implicit none
    
    integer, parameter :: FIRST_CELL = 1
    integer, parameter :: SECOND_CELL = 2
    integer, parameter :: NO_CELL = -1
    integer, parameter :: MAX_SIZE = 1000
    
    logical :: wallsUp ( MAX_SIZE )
    integer :: wallConnections ( MAX_SIZE, 2 )
    integer :: cellToGroup ( MAX_SIZE )
    integer :: groupCells ( MAX_SIZE, MAX_SIZE )
    integer :: wallRemoveList ( MAX_SIZE )
    
    integer :: mazeColumns, mazeRows, interiorWallCount
    integer :: wallIndex, rowIndex, columnIndex
    integer :: firstCellInRow, leftCell, rightCell
    integer :: upperCell, lowerCell, cellIndex
    integer :: innerCellIndex, shuffleIndex, otherIndex
    integer :: removeWallIndex, nextWallToCheck
    integer :: firstCell, firstCellGroupIndex
    integer :: secondCell, secondCellGroupIndex
    integer :: nextEmptyFirstGroupIndex
    integer :: groupCellIndex, cellToMove
    integer :: temp, ios, seed
    logical :: mazeComplete
    character ( len=100 ) :: inputString
    
    ! Prompt for maze size
    mazeColumns = 0
    do while ( mazeColumns <= 0 )
        write ( *, '(A)', advance='no' ) 'Please enter number of columns for maze, must be greater than 1: '
        read ( *, '(A)', iostat=ios ) inputString
        if ( ios == 0 ) then
            read ( inputString, *, iostat=ios ) mazeColumns
            if ( ios /= 0 ) mazeColumns = 0
        end if
    end do
    
    mazeRows = 0
    do while ( mazeRows <= 0 )
        write ( *, '(A)', advance='no' ) 'Please enter number of rows for maze, must be greater than 1: '
        read ( *, '(A)', iostat=ios ) inputString
        if ( ios == 0 ) then
            read ( inputString, *, iostat=ios ) mazeRows
            if ( ios /= 0 ) mazeRows = 0
        end if
    end do
    
    interiorWallCount = mazeRows * ( mazeColumns - 1 ) + ( mazeRows - 1 ) * mazeColumns
    
    ! Initialize walls (all up)
    do wallIndex = 1, interiorWallCount
        wallsUp ( wallIndex ) = .true.
    end do
    
    ! Build wall connections
    wallIndex = 1
    do rowIndex = 0, mazeRows - 1
        firstCellInRow = rowIndex * mazeColumns + 1
        
        ! Vertical walls
        do columnIndex = 0, mazeColumns - 2
            leftCell = firstCellInRow + columnIndex
            rightCell = leftCell + 1
            wallConnections ( wallIndex, FIRST_CELL ) = leftCell
            wallConnections ( wallIndex, SECOND_CELL ) = rightCell
            wallIndex = wallIndex + 1
        end do
        
        ! Horizontal walls
        if ( wallIndex <= interiorWallCount ) then
            do columnIndex = 0, mazeColumns - 1
                upperCell = firstCellInRow + columnIndex
                lowerCell = upperCell + mazeColumns
                wallConnections ( wallIndex, FIRST_CELL ) = upperCell
                wallConnections ( wallIndex, SECOND_CELL ) = lowerCell
                wallIndex = wallIndex + 1
            end do
        end if
    end do
    
    ! Initialize cell groups
    do cellIndex = 1, mazeRows * mazeColumns
        cellToGroup ( cellIndex ) = cellIndex
    end do
    
    ! Initialize group cells
    do cellIndex = 1, mazeRows * mazeColumns
        do innerCellIndex = 1, mazeRows * mazeColumns
            if ( innerCellIndex == 1 ) then
                groupCells ( cellIndex, innerCellIndex ) = cellIndex
            else
                groupCells ( cellIndex, innerCellIndex ) = NO_CELL
            end if
        end do
    end do
    
    ! Print initial maze
    call printMaze ( )
    
    ! Initialize wall removal list
    do wallIndex = 1, interiorWallCount
        wallRemoveList ( wallIndex ) = wallIndex
    end do
    
    ! Initialize random seed
    call random_seed ( )
    seed = 12345
    
    ! Fisher-Yates shuffle algorithm
    do shuffleIndex = interiorWallCount, 2, -1
        ! Simple LCG for random number generation
        seed = mod ( seed * 1664525 + 1013904223, 2147483647 )
        otherIndex = mod ( abs ( seed ), shuffleIndex ) + 1
        temp = wallRemoveList ( shuffleIndex )
        wallRemoveList ( shuffleIndex ) = wallRemoveList ( otherIndex )
        wallRemoveList ( otherIndex ) = temp
    end do
    
    mazeComplete = .false.
    
    ! Kruskal's algorithm
    do removeWallIndex = 1, interiorWallCount
        nextWallToCheck = wallRemoveList ( removeWallIndex )
        
        firstCell = wallConnections ( nextWallToCheck, FIRST_CELL )
        firstCellGroupIndex = cellToGroup ( firstCell )
        secondCell = wallConnections ( nextWallToCheck, SECOND_CELL )
        secondCellGroupIndex = cellToGroup ( secondCell )
        
        if ( firstCellGroupIndex /= secondCellGroupIndex ) then
            wallsUp ( nextWallToCheck ) = .false.
            
            ! Find next empty position in first group
            nextEmptyFirstGroupIndex = 1
            do cellIndex = 1, mazeColumns * mazeRows
                if ( groupCells ( firstCellGroupIndex, cellIndex ) == NO_CELL ) then
                    nextEmptyFirstGroupIndex = cellIndex
                    exit
                end if
            end do
            
            ! Move all cells from second group to first group
            do groupCellIndex = mazeColumns * mazeRows, 1, -1
                if ( groupCells ( secondCellGroupIndex, groupCellIndex ) /= NO_CELL ) then
                    cellToMove = groupCells ( secondCellGroupIndex, groupCellIndex )
                    
                    groupCells ( firstCellGroupIndex, nextEmptyFirstGroupIndex ) = cellToMove
                    nextEmptyFirstGroupIndex = nextEmptyFirstGroupIndex + 1
                    cellToGroup ( cellToMove ) = firstCellGroupIndex
                    groupCells ( secondCellGroupIndex, groupCellIndex ) = NO_CELL
                    
                    if ( nextEmptyFirstGroupIndex > mazeColumns * mazeRows ) then
                        mazeComplete = .true.
                    end if
                end if
            end do
            
            ! Sleep and print maze
            call sleep ( 1 )
            write ( *, * )
            call printMaze ( )
            
            if ( mazeComplete ) exit
        end if
    end do
    
contains
    
    subroutine printMaze ( )
        integer :: currentInteriorWall, rowIdx, colIdx
        
        currentInteriorWall = 1
        
        ! Print top border
        write ( *, '(A)', advance='no' ) '+'
        do colIdx = 1, mazeColumns
            write ( *, '(A)', advance='no' ) '-+'
        end do
        write ( *, * )
        
        do rowIdx = 1, mazeRows
            ! Print vertical walls and cells
            write ( *, '(A)', advance='no' ) '|'
            do colIdx = 1, mazeColumns
                write ( *, '(A)', advance='no' ) ' '
                
                if ( colIdx == mazeColumns .or. wallsUp ( currentInteriorWall ) ) then
                    write ( *, '(A)', advance='no' ) '|'
                else
                    write ( *, '(A)', advance='no' ) ' '
                end if
                
                if ( colIdx < mazeColumns ) then
                    currentInteriorWall = currentInteriorWall + 1
                end if
            end do
            write ( *, * )
            
            ! Print horizontal walls
            if ( rowIdx < mazeRows ) then
                write ( *, '(A)', advance='no' ) '+'
                do colIdx = 1, mazeColumns
                    if ( wallsUp ( currentInteriorWall ) ) then
                        write ( *, '(A)', advance='no' ) '-'
                    else
                        write ( *, '(A)', advance='no' ) ' '
                    end if
                    write ( *, '(A)', advance='no' ) '+'
                    currentInteriorWall = currentInteriorWall + 1
                end do
                write ( *, * )
            end if
        end do
        
        ! Print bottom border
        write ( *, '(A)', advance='no' ) '+'
        do colIdx = 1, mazeColumns
            write ( *, '(A)', advance='no' ) '-+'
        end do
        write ( *, * )
    end subroutine printMaze
    
end program mazeGenerator
